### PR TITLE
notices: ensure that two notices don't have the same lastDateTime

### PR DESCRIPTION
**Pull request by sergio-costas**
When a client wants to receive the notices using the /v2/notices interface, it usually will first receive all the old notifications, store the date and time of the last notification, and in the next calls use that date and time to get the next notices, and avoid getting the old notices over and over again.

Unfortunately, if, due to chance, two or more notifications have the same date and time, this scheme will fail, because after receiving the first one, the call will return it, and when the client asks for the next, it will filter from the date/time of the last notice sent, so the second notice with the same timestamp won't be sent. This can happen if the timer used for timestamps doesn't have enough granularity.

This patch fixes this by ensuring that no notice has the same date/time value for lastReceived, by adding one or more nanoseconds when required.

Thanks for helping us make a better snapd!
Have you signed the [license agreement](https://www.ubuntu.com/legal/contributors) and read the [contribution guide](https://github.com/snapcore/snapd/blob/master/CONTRIBUTING.md)?

### CODE REVIEW

**Comment by sergio-costas** [comment]
@ernestl Is there a folder for "utility functions" where to put the CompareDates() function? Maybe inside timeutil?

File: overlord/state/notices.go [hunk 1]
```diff
@@ -243,7 +259,12 @@ func (s *State) AddNotice(userID *uint32, noticeType NoticeType, key string, opt
 
 	now := options.Time
 	if now.IsZero() {
-		now = time.Now()
+		now = getTimeNow()
```
Comment (ZeyadYasser) [inline overlord/state/notices.go:262]: `time.Now()` is already mock-able.
```suggestion
		now = timeNow()
```

File: overlord/state/notices.go [hunk 2]
```diff
@@ -243,7 +259,12 @@ func (s *State) AddNotice(userID *uint32, noticeType NoticeType, key string, opt
...
+		// ensure that two notices never have the same sent time
+		if compareDates(now, s.noticeLastDate) <= 0 {
```
Comment (ZeyadYasser) [inline overlord/state/notices.go:264]: do we need the time compare helper? wouldn't the check below be sufficient?
```suggestion
		if !now.After(s.noticeLastDate) {
```

File: overlord/state/state.go [hunk 3]
```diff
@@ -92,12 +92,13 @@ type State struct {
...
-	tasks    map[string]*Task
-	warnings map[string]*Warning
-	notices  map[noticeKey]*Notice
+	backend        Backend
+	data           customData
+	changes        map[string]*Change
+	tasks          map[string]*Task
+	warnings       map[string]*Warning
+	notices        map[noticeKey]*Notice
+	noticeLastDate time.Time
```
Comment (ZeyadYasser) [inline overlord/state/state.go:101]: I believe this needs to persisted, please check `State.MarshalJSON()` and `State.UnmarshalJSON()`.

**Review by ZeyadYasser — CHANGES_REQUESTED** [review]
Thank you! I did a first pass.

File: overlord/state/notices.go (hunk 1 above)
Comment (sergio-costas) [reply overlord/state/notices.go:262]: Good point. Changing everything.

File: overlord/state/notices.go [hunk 4]
```diff
@@ -243,7 +259,12 @@ func (s *State) AddNotice(userID *uint32, noticeType NoticeType, key string, opt
 
 	now := options.Time
 	if now.IsZero() {
-		now = time.Now()
+		now = getTimeNow()
+		// ensure that two notices never have the same sent time
+		if !now.After(s.noticeLastDate) {
+			now = s.noticeLastDate.Add(time.Nanosecond)
```
Comment (pedronis) [inline overlord/state/notices.go:265]: I'm confused, does parseOptionalTime in daemon support ns precision?

Comment (sergio-costas) [reply overlord/state/notices.go:265]: Yes. In fact, I added an specific test for that. Is the next PR.

Comment (sergio-costas) [reply overlord/state/notices.go:249]: The microsecond problem is only in GLib. GDateTime internally limits everything to microseconds. That confused me.

File: overlord/state/export_test.go [hunk 5]
```diff
@@ -72,3 +72,15 @@ var (
 func (s *State) NumNotices() int {
 	return len(s.notices)
 }
+
+func MockGetTimeNow(f func() time.Time) (restore func()) {
```
Comment (ZeyadYasser) [inline overlord/state/export_test.go:76]: I think this can be removed

File: overlord/state/export_test.go [hunk 6]
```diff
@@ -72,3 +72,15 @@ var (
...
+	old := timeNow
+	timeNow = f
+	return func() {
+		timeNow = old
+	}
+}
+
+func (n *Notice) GetLastOccurred() time.Time {
```
Comment (ZeyadYasser) [inline overlord/state/export_test.go:84]: No need to expose this for tests, I believe notices tests use `noticeToMap` to get unexported notice data.

File: overlord/state/state.go (hunk 3 above)
Comment (ZeyadYasser) [reply overlord/state/state.go:101]: Good point, maybe we don't need to persist it.

CC: @pedronis

**Review by ZeyadYasser — COMMENTED** [review]
Thank you for the changes! I did a second pass

File: overlord/state/export_test.go (hunk 5 above)
Comment (sergio-costas) [reply overlord/state/export_test.go:76]: I tried, and changed the calls to `MockTimeNow()`, but it doesn't work. There's no mock function to mock timenow in the `state` package. Anyway, I renamed it to MockTimeNow for consistency.

File: overlord/state/export_test.go (hunk 6 above)
Comment (sergio-costas) [reply overlord/state/export_test.go:84]: The problem is that `noticeToMap()` uses JSON marshalling to do its job, and one of the tests is to ensure that the dates are correct even after marshalling/unmarshalling...

File: overlord/state/state.go (hunk 3 above)
Comment (ernestl) [reply overlord/state/state.go:101]: My view is its dangerous to rely on time elapsing/delays to guarantee correct behaviour. I would play it safe unless there is massive downside to consider.

File: overlord/state/export_test.go (hunk 5 above)
Comment (ZeyadYasser) [reply overlord/state/export_test.go:76]: I believe that in the `state` package it is called `MockTime`

File: overlord/state/state.go (hunk 3 above)
Comment (sergio-costas) [reply overlord/state/state.go:101]: I agree. So, since persistence is already implemented, I'll close this.

**Comment by olivercalder** [comment]
CC @benhoyt

File: overlord/state/export_test.go (hunk 6 above)
Comment (benhoyt) [reply overlord/state/export_test.go:76]: I agree with @ZeyadYasser. We're not testing JSON marshalling here, and JSON marshalling includes full nanoseconds (the resolution of `time.Duration`). So I think we should drop this and test via `noticeToMap` like everything else.

File: overlord/state/notices.go [hunk 7]
```diff
@@ -243,7 +243,12 @@ func (s *State) AddNotice(userID *uint32, noticeType NoticeType, key string, opt
 
 	now := options.Time
 	if now.IsZero() {
-		now = time.Now()
+		now = timeNow()
+		// ensure that two notices never have the same sent time
+		if !now.After(s.noticeLastDate) {
```
Comment (benhoyt) [inline overlord/state/notices.go:248]: Given we're comparing equality, why not `if now.Equal(s.noticeLastDate)`?

File: overlord/state/state.go (hunk 3 above)
Comment (benhoyt) [inline overlord/state/state.go:101]: I think this should be named `noticeLastOccurred` as occurred / last occurred is what that field is called. Similar for the JSON field.

File: overlord/state/notices_test.go [hunk 8]
```diff
@@ -679,6 +679,93 @@ func (s *noticesSuite) TestValidateNotice(c *C) {
 	c.Check(id, Equals, "")
 }
 
+func (s *noticesSuite) TestAvoidTwoNoticesWithSameDateTime(c *C) {
```
Comment (benhoyt) [inline overlord/state/notices_test.go:682]: It seems to me this test is testing too much. All we need to test is that two notices with the same (fake) timeNow end up with notice2.lastOccurred > notice1.lastOccurred, right?

File: overlord/state/state.go [hunk 9]
```diff
@@ -171,6 +172,8 @@ type marshalledState struct {
 	LastTaskId   int `json:"last-task-id"`
 	LastLaneId   int `json:"last-lane-id"`
 	LastNoticeId int `json:"last-notice-id"`
+
+	NoticeLastDate time.Time `json:"notice-last-date,omitempty"`
```
Comment (olivercalder) [inline overlord/state/state.go:176]: Is there any way to get json to marshal this using a resolution more precise than 1 second? Perhaps a custom {,un}marshaller which uses `RFC3339Nano` would be best?

Comment (olivercalder) [reply overlord/state/state.go:176]: In particular, what if this were a string...

File: overlord/state/state.go [hunk 10]
```diff
@@ -187,6 +190,8 @@ func (s *State) MarshalJSON() ([]byte, error) {
 		LastChangeId: s.lastChangeId,
 		LastLaneId:   s.lastLaneId,
 		LastNoticeId: s.lastNoticeId,
+
+		NoticeLastDate: s.noticeLastDate,
```
Comment (olivercalder) [inline overlord/state/state.go:194]: ...and this were:
```go
NoticeLastDate: s.noticeLastDate.Format(time.RFC3339Nano)
```

File: overlord/state/state.go [hunk 11]
```diff
@@ -207,6 +212,7 @@ func (s *State) UnmarshalJSON(data []byte) error {
 	s.lastTaskId = unmarshalled.LastTaskId
 	s.lastLaneId = unmarshalled.LastLaneId
 	s.lastNoticeId = unmarshalled.LastNoticeId
+	s.noticeLastDate = unmarshalled.NoticeLastDate
```
Comment (olivercalder) [inline overlord/state/state.go:215]: ...and this were:
```go
s.noticeLastDate = time.Parse(time.RFC3339Nano, unmarshalled.NoticeLastDate)
```

**Review by olivercalder — COMMENTED** [review]
Overall looks good, but I think persisting with 1-second granularity does much good. Persisting with RFC3339Nano format would be preferable, in my opinion. Or even unix nano, if that's a thing in Go (I don't see it [here](https://pkg.go.dev/time#pkg-constants)).

Also, nitpick, but my personal preference would be to use `[Nn]oticeLastTimestamp` instead of `[Nn]oticeLastDate`.

**Review by benhoyt — APPROVED** [review]
I've left a few minor comments.

I'm curious how this came up, or is it a purely theoretical problem?

As the original implementer of notices (in [Pebble PR 295](https://github.com/canonical/pebble/pull/295)), I originally told myself this wasn't necessary, because `time.Time` is nanosecond precision ... but of course `time.Now` is not guaranteed to have that resolution on a given system. For what it's worth, on my Linux amd64 laptop the smallest delta between `time.Now` calls I see is 22ns, and between two `state.AddNotice` calls about 200ns.

But I can see why this is a good idea, and it's fairly trivial to do -- thanks.

File: overlord/state/state.go (hunk 9 above)
Comment (benhoyt) [reply overlord/state/state.go:176]: Go's `encoding/json` does use nanosecond precision for these: https://go.dev/play/p/nmJ_2l5G6QI

**Comment by benhoyt** [comment]
Sorry, I didn't intend to click Approve there, just comment. I realise I'm not an approver around here :-), though I would want those changes I suggested for backporting this to Pebble.

File: overlord/state/notices.go (hunk 7 above)
Comment (olivercalder) [reply overlord/state/notices.go:248]: Actually, I think using `!now.After(s.noticeLastDate)` would also solve the non-monotonic time problem. It might result in notices being created in the "future" according to current system time after a synchronization results in system time moving backwards, but after $/Delta$ seconds, everything would be in order and no notice could ever be issued "before" an existing notice.

Comment (sergio-costas) [reply overlord/state/notices.go:248]: Using equality would fail if there are three notices with the same timestamp. The first one would have `now()` value, and that value would be stored for future comparisons; the next one, also with `now()`, would be equal to the stored value, so `now()+1` would be used and the stored value would be updated. But the third one would also have `now()`, which is smaller than the value stored, so the algorithm would fail. Using `!After()` ensures that it doesn't matter how many notices have the same timestamp.

**Comment by sergio-costas** [comment]
@benhoyt Well, this came up due to a resolution problem that I assumed was in Go, but in the end it was in the `snapd-glib` API that I was using to communicate with (GDateTime has one microsecond resolution). But after that, I began to read about timers in the PCs and found an interesting rabbit hole.

TL;DR although since the original Pentium there are timers with very high resolution, and today any recent X86_64 device has timers with nanosecond granularity, this can't be guaranteed with not-so-recent devices or in other architectures, where, under some circumstances, granularity can be as low as tenths or hundredths of microsecond.

The original Pentium added counters that incremented with the processor's clock, so in the original Pentium at 50MHz they resolved to 200ns. Higher clock speeds result in better granularity, and any processor with 1GHz or faster clock would have nanosecond granularity. But when Power Management was added, the clock could vary or even be stopped, having to rely on the HPET, with clock frequency of about 14MHz according to Wikipedia (thus, about 71ns). This was fixed in later models, but then multicore devices arrived, and in those, these counters weren't guaranteed to be synchronized between cores, so software solutions were needed (like a daemon for AMD K8 processors that periodically re-synchronized the counters between cores). Most modern x86_64 processors fixed this problem too, so currently we can have one nanosecond granularity in the timers. ARM seems to have similar counters, but I haven't found much info about them. And, anyway, there are still devices like the Raspberry Pi 3 with 400MHz clock speed that, if I get it correctly, couldn't have better granularity than 2.5ns.

So, since we can't guarantee that we always have 1 nanosecond granularity in the system clock, I think that it's a good idea to add this patch.

Sources:
https://learn.microsoft.com/en-us/windows/win32/sysinfo/acquiring-high-resolution-time-stamps
https://en.wikipedia.org/wiki/Time_Stamp_Counter
https://0xax.gitbooks.io/linux-insides/content/Timers/linux-timers-6.html
https://forums.raspberrypi.com/viewtopic.php?t=133981
https://lkml.org/lkml/2005/11/4/173

File: overlord/state/notices_test.go (hunk 8 above)
Comment (sergio-costas) [reply overlord/state/notices_test.go:682]: It's better to test three, to avoid the error that could result in using `Equal()` instead of `!After()`.

File: overlord/state/state.go (hunk 11 above)
Comment (sergio-costas) [reply overlord/state/state.go]: Done.

File: overlord/state/state.go (hunk 10 above)
Comment (sergio-costas) [reply overlord/state/state.go]: Done.

File: overlord/state/notices_test.go (hunk 8 above)
Comment (sergio-costas) [reply overlord/state/notices_test.go:682]: Done. Now it uses `noticeToMap()`.

File: overlord/state/export_test.go (hunk 6 above)
Comment (sergio-costas) [reply overlord/state/export_test.go:76]: Done.

File: cmd/snap-bootstrap/cmd_initramfs_mounts_test.go [hunk 12]
```diff
@@ -3310,7 +3310,7 @@ grade=signed
 		c.Check(fiParent.Mode(), Equals, os.FileMode(os.ModeDir|0750))
 	}
 
-	c.Check(filepath.Join(ephemeralUbuntuData, "system-data/var/lib/snapd/state.json"), testutil.FileEquals, `{"data":{"auth":{"last-id":1,"macaroon-key":"not-a-cookie","users":[{"id":1,"name":"mvo"}]}},"changes":{},"tasks":{},"last-change-id":0,"last-task-id":0,"last-lane-id":0,"last-notice-id":0}`)
+	c.Check(filepath.Join(ephemeralUbuntuData, "system-data/var/lib/snapd/state.json"), testutil.FileEquals, `{"data":{"auth":{"last-id":1,"macaroon-key":"not-a-cookie","users":[{"id":1,"name":"mvo"}]}},"changes":{},"tasks":{},"last-change-id":0,"last-task-id":0,"last-lane-id":0,"last-notice-id":0,"notice-last-date":"0001-01-01T00:00:00Z"}`)
```
Comment (olivercalder) [inline cmd/snap-bootstrap/cmd_initramfs_mounts_test.go:3313]: I believe this should be `"notice-last-timestamp"` now. Also, if golang json encodings of times are nanosecond precision, shouldn't the test string include nanosecond precision here as well?

File: overlord/overlord_test.go [hunk 13]
```diff
@@ -176,7 +176,8 @@ func (ovs *overlordSuite) TestNewWithGoodState(c *C) {
 		"last-change-id": 0,
 		"last-task-id": 0,
 		"last-lane-id": 0,
-		"last-notice-id": 0
+		"last-notice-id": 0,
+		"notice-last-date":"0001-01-01T00:00:00Z"
```
Comment (olivercalder) [inline overlord/overlord_test.go:180]: Here as well: `"notice-last-timestamp"` and nanosecond precision?

File: overlord/state/copy_test.go [hunk 14]
```diff
@@ -79,7 +79,7 @@ var srcStateContent = []byte(`
 }
 `)
 
-const stateSuffix = `,"changes":{},"tasks":{},"last-change-id":0,"last-task-id":0,"last-lane-id":0,"last-notice-id":0}`
+const stateSuffix = `,"changes":{},"tasks":{},"last-change-id":0,"last-task-id":0,"last-lane-id":0,"last-notice-id":0,"notice-last-date":"0001-01-01T00:00:00Z"}`
```
Comment (olivercalder) [inline overlord/state/copy_test.go:82]: Same

File: overlord/state/state.go (hunk 9 above)
Comment (olivercalder) [reply overlord/state/state.go:176]: Ah okay. Then I think the test strings should use nanosecond precision as well, right? I commented on the locations.

Comment (olivercalder) [inline overlord/state/state.go:176]: Should this be changed to `NoticeLastTimestamp` with `"notice-last-timestamp"` as well?

File: overlord/state/state.go [hunk 15]
```diff
@@ -187,6 +190,8 @@ func (s *State) MarshalJSON() ([]byte, error) {
 		LastChangeId: s.lastChangeId,
 		LastLaneId:   s.lastLaneId,
 		LastNoticeId: s.lastNoticeId,
+
+		NoticeLastDate: s.noticeLastTimestamp,
```
Comment (olivercalder) [inline overlord/state/state.go:194]: I think `[Nn]oticeLastTimestamp` for both?

File: overlord/state/state.go [hunk 16]
```diff
@@ -207,6 +212,7 @@ func (s *State) UnmarshalJSON(data []byte) error {
 	s.lastTaskId = unmarshalled.LastTaskId
 	s.lastLaneId = unmarshalled.LastLaneId
 	s.lastNoticeId = unmarshalled.LastNoticeId
+	s.noticeLastTimestamp = unmarshalled.NoticeLastDate
```
Comment (olivercalder) [inline overlord/state/state.go:215]: Same

**Review by olivercalder — COMMENTED** [review]
Thanks! A few notes on consistency with "LastDate" vs "LastTimestamp" and using nanosecond precision in json strings.

File: cmd/snap-bootstrap/cmd_initramfs_mounts_test.go (hunk 12 above)
Comment (sergio-costas) [reply cmd/snap-bootstrap/cmd_initramfs_mounts_test.go:3313]: Ops! I was filtering `*test.go` files, so I didn't find them. Sorry. Fixing all!

File: overlord/state/copy_test.go (hunk 14 above)
Comment (sergio-costas) [reply overlord/state/copy_test.go:82]: Fixed too.

File: overlord/state/state.go (hunk 9 above)
Comment (sergio-costas) [reply overlord/state/state.go:176]: Yes. Done.

File: overlord/overlord_test.go (hunk 13 above)
Comment (sergio-costas) [reply overlord/overlord_test.go:180]: AFAIK, nanosecond precision is really not needed if it's zero...

File: overlord/state/state.go (hunk 16 above)
Comment (sergio-costas) [reply overlord/state/state.go]: done.

File: overlord/overlord_test.go (hunk 13 above)
Comment (sergio-costas) [reply overlord/overlord_test.go]: Ok, tested, and confirmed that it removes trailing zeros, so an integer value (zero in this case) won't have "dot zero zero zero..."

File: overlord/state/state.go (hunk 9 above)
Comment (sergio-costas) [reply overlord/state/state.go]: As commented below, it seems that the trailing zeros are removed, so for the tests to pass, they must not have decimals.

File: overlord/overlord_test.go (hunk 13 above)
Comment (olivercalder) [reply overlord/overlord_test.go]: My thought was to test that everything works as expected if the timestamp has nonzero nanoseconds, since afaik the odds of the nanoseconds being zero in the real world are one in a billion :)

Am I missing something though, is there a reason the time will be integer-valued seconds?

File: overlord/state/notices.go (hunk 7 above)
Comment (olivercalder) [reply overlord/state/notices.go]: @zyga would this address your concerns with relying on notice timestamps?

File: overlord/state/export_test.go [hunk 17]
```diff
@@ -72,3 +72,7 @@ var (
 func (s *State) NumNotices() int {
 	return len(s.notices)
 }
+
+func (n *Notice) GetLastOccurred() time.Time {
```
Comment (benhoyt) [inline overlord/state/export_test.go:76]: We should drop this method now that it's not needed, right?

File: overlord/overlord_test.go (hunk 13 above)
Comment (benhoyt) [reply overlord/overlord_test.go]: [Playground link](https://go.dev/play/p/LJ-5ojPjXbI) that demonstrates how Go's encoding/json removes trailing zeroes:

```
time.Date(2024, time.January, 2, 3, 4, 5, 0, time.UTC) -> "2024-01-02T03:04:05Z"
time.Date(2024, time.January, 2, 3, 4, 5, 100000000, time.UTC) -> "2024-01-02T03:04:05.1Z"
time.Date(2024, time.January, 2, 3, 4, 5, 120000000, time.UTC) -> "2024-01-02T03:04:05.12Z"
time.Date(2024, time.January, 2, 3, 4, 5, 123000000, time.UTC) -> "2024-01-02T03:04:05.123Z"
time.Date(2024, time.January, 2, 3, 4, 5, 123400000, time.UTC) -> "2024-01-02T03:04:05.1234Z"
time.Date(2024, time.January, 2, 3, 4, 5, 123450000, time.UTC) -> "2024-01-02T03:04:05.12345Z"
time.Date(2024, time.January, 2, 3, 4, 5, 123456000, time.UTC) -> "2024-01-02T03:04:05.123456Z"
time.Date(2024, time.January, 2, 3, 4, 5, 123456700, time.UTC) -> "2024-01-02T03:04:05.1234567Z"
time.Date(2024, time.January, 2, 3, 4, 5, 123456780, time.UTC) -> "2024-01-02T03:04:05.12345678Z"
time.Date(2024, time.January, 2, 3, 4, 5, 123456789, time.UTC) -> "2024-01-02T03:04:05.123456789Z"
```

So the JSON-encoded zero value of time.Time will not have nanoseconds. It looks like we do use nanoseconds in the test, so I think this is fine.

**Review by benhoyt — COMMENTED** [review]
Looks good to me, thanks for your insights here and the write-up about time resolution -- that was very interesting!

File: overlord/overlord_test.go (hunk 13 above)
Comment (olivercalder) [reply overlord/overlord_test.go]: Apologies, I didn't really process that this was encoding the zero-value for the time type, not an explicitly-specified time, so I see why we can't have nanosecond precision here. And the new `TestAvoidTwoNoticesWithSameDateTime` checks json marshalling/unmarshalling with nonzero nanoseconds. So all good from me, thanks for your patience :)

File: overlord/state/export_test.go (hunk 17 above)
Comment (sergio-costas) [reply overlord/state/export_test.go:76]: Ops!!! You are right. Fixing!

**Review by olivercalder — APPROVED** [review]
LTGM, thank you!

@zyga I do still think this (for the most part) solves the monotonic time problem. If the server time goes backwards and a new notice is added, this change guarantees the new notice will occur with a timestamp later than all previously-recorded notices. Thus, with any "after" filter based on an existing notice timestamp (as is the general use-case for timestamps in notices), the newly-issued notice will still be included in the response.

From the `AddNotice` side of things, where non-monotonic time might result in a missed signal due to the `RepeatAfter` value being incorrect, the `now` value which is compared against existing notices' `lastRepeated` value is adjusted to always be after any previous notice timestamp. So time can never really go backwards, it just may pause for a moment until the wall clock again passes the `noticeLastTimestamp`. So I suppose notices recorded while time is paused could get lost, but afaik this would only happen if the time differential is larger than the `RepeatAfter` option, and this seems unlikely. It would be problematic though if this were to occur.

So I've convinced myself that it's not fully solved, though this feels like an improvement :)

**Comment by sergio-costas** [comment]
@olivercalder Mmm... but why would the time in the server "go backwards" by such a big time? I mean: if you are talking about daylight saving, the timezone is also changed, so the nanoseconds will still be monotonically increasing because they should be calculated against UTC, by adding/subtracting the timezone to the local time. The only case where the time goes really backwards is if the internal clock has some drift (which is inevitable), and NTP adjusts it. But that difference would be milliseconds at most, so the `RepeatAfter` should not be a problem...

**Comment by olivercalder** [comment]
> @olivercalder Mmm... but why would the time in the server "go backwards" by such a big time? I mean: if you are talking about daylight saving, the timezone is also changed, so the nanoseconds will still be monotonically increasing because they should be calculated against UTC, by adding/subtracting the timezone to the local time. The only case where the time goes really backwards is if the internal clock has some drift (which is inevitable), and NTP adjusts it. But that difference would be milliseconds at most, so the `RepeatAfter` should not be a problem...

@sergio-costas I believe the problem would be if the `RepeatAfter` time is on the order of the NTP adjustments. I agree this is unlikely, and it's especially unlikely that a notice which one wants to assure will not be missed would have a `RepeatAfter` time set at all, since that by definition would have the potential to suppress essential messages if they occur within the `RepeatAfter` timeframe. So I think functionally this shouldn't be a problem in practice, but technically it is possible for it to occur, which @zyga pointed out to me in a discussion we had a few weeks ago. AFAICT though, the problem of lost signals only remains if one uses `RepeatAfter` ...?

File: overlord/state/state.go [hunk 18]
```diff
@@ -92,12 +92,13 @@ type State struct {
 	// for registering runtime callbacks
 	lastHandlerId int
```
Comment (ZeyadYasser) [inline overlord/state/state.go:93]: Small nitpick, I believe it would be better to move it up here and `s/noticeLastTimestamp/lastNoticeTimestamp`
```suggestion
	lastHandlerId int
	lastNoticeTimestamp time.Time
```

**Review by ZeyadYasser — APPROVED** [review]
Great work, Great catch, Thank you!

File: overlord/state/state.go (hunk 18 above)
Comment (sergio-costas) [reply overlord/state/state.go]: Agree. Done.

**Comment by sergio-costas** [comment]
Rebased.

**Review by ernestl — APPROVED** [review]

**Review by pedronis — APPROVED** [review]

**Review by ernestl — APPROVED** [review]
Looks good after final commit with explanation.
